keps/sig-multicluster/2149-clusterid/README.md

```diff
@@ -0,0 +1,705 @@
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Overview
+Each cluster in a ClusterSet will be assigned a unique identifier that lives at least as long as its membership in the set, and is immutable for its lifetime. This identifier will be stored in a new cluster-scoped `ClusterClaim` CR with the well known name `id.k8s.io` that may be referenced by workloads within the cluster. The identifier must be a valid [RFC-1123](https://tools.ietf.org/html/rfc1123) DNS label, and may be created by an implementation dependent mechanism.
+
+While a member of a ClusterSet, a cluster will also have an additional `clusterset.k8s.io ClusterClaim` which describes its current membership. This claim must be present exactly as long as the set membership lasts, and removed when the cluster is no longer a member.
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### ClusterSet membership
+
+I have some set of clusters working together and need a way to uniquely identify them within the system that I use to track membership.
+
+#### Multi-Cluster Services
+
+I have a headless multi-cluster service deployed across clusters in my ClusterSet with similarly named pods in each cluster. I need a way to disambiguate each backend pod via DNS.
+
+#### Diagnostics
+
+Clusters within my ClusterSet send logs/metrics to a common monitoring solution and I need to be able to identify the cluster from which a given set of events originated.
+
+
+#### Joining or moving between ClusterSets
+
+I want the ability to add a previously-isolated cluster to a ClusterSet, or to move a cluster from one ClusterSet to another and be aware of this change.
+
+### `ClusterClaim` CRD
+
+The `ClusterClaim` resource provides a way to store identification related, cluster scoped information for multi-cluster tools while creating flexibility for implementations. A cluster may have multiple `ClusterClaim`s, each holding a different identification related value. Each claim contains the following information:
+
+*   **Name** - a well known or custom name to identify the claim.
+*   **Value** - a claim-dependent string.
+
+
+### Well known claims
+
+The requirements below use the keywords **must, should,** and **may** purposefully in accordance with [RFC-2119](https://tools.ietf.org/html/rfc2119).
+
+
+#### Claim: `id.k8s.io`
+
+Contains a unique identifier for the containing cluster.
+
+
+##### Uniqueness
+
+*   The identifier **must** be unique within the ClusterSet to which its cluster belongs for the duration of the cluster’s membership.
```

Comment:
How will we implement this constraint? I don't think we can ensure uniqueness at the level of the individual clusters. Will we instead implement it by enforcing this invariant at the level of the ClusterSet?

Reply:
I think this would have to be implemented at the ClusterSet level. Whatever manages memberships could e.g. deny a cluster from joining if an existing member has the specified ID, or perhaps assign a new one

Reply:
Added some notes from this thread in 36dc62a0aab462665b324f80ae73e01475021336 as seeds for the `implementable` review round